tests/bsim/bluetooth/mesh/src/test_beacon.c

```diff
@@ -908,6 +971,539 @@ static void test_rx_secure_beacon_interval(void)
 	PASS();
 }
 
+#if CONFIG_BT_MESH_PRIV_BEACONS
+static void scan_recv(const struct bt_le_scan_recv_info *info, struct net_buf_simple *buf)
+{
+	int64_t beacon_time_delta;
+
+	/* Strip length field */
+	(void)net_buf_simple_pull_u8(buf);
+	uint8_t type = net_buf_simple_pull_u8(buf);
+
+	if (type == BT_DATA_MESH_BEACON && net_buf_simple_pull_u8(buf) == BEACON_TYPE_PRIVATE) {
+		memcpy(curr_random, buf->data, 13);
+		bt_addr_le_copy(&curr_beacon_adv_addr, info->addr);
+
+		if (last_beacon_time == 0) {
+			LOG_DBG("First Private Beacon received");
+		} else {
+			beacon_time_delta = k_uptime_delta(&last_beacon_time);
+			LOG_DBG("Private Beacon received after %dms", beacon_time_delta);
+			beaconing_time += beacon_time_delta;
+		}
+
+		last_beacon_time = k_uptime_get();
+		k_sem_give(&prb_received);
+	}
+}
+
+static struct bt_le_scan_cb scan_callbacks = {
+	.recv = scan_recv,
+};
+
+static void provision(uint16_t addr)
+{
+	int err;
+
+	err = bt_mesh_provision(net_key, 0, 0, 0, addr, dev_key);
+	if (err) {
+		FAIL("Provisioning failed (err %d)", err);
+		return;
+	}
+}
+
+static void common_configure(uint16_t addr)
+{
+	uint8_t status;
+	int err;
+
+	err = bt_mesh_cfg_cli_app_key_add(0, addr, 0, 0, app_key, &status);
+	if (err || status) {
+		FAIL("AppKey add failed (err %d, status %u)", err, status);
+		return;
+	}
+}
+
+/* Since nodes self-provision in this test and the Private Beacon model uses device keys for crypto,
+ * the server node must be added to the client CDB manually.
+ */
+static void target_node_alloc(struct bt_mesh_comp comp, uint16_t addr, const uint8_t *dev_key)
+{
+	struct bt_mesh_cdb_node *node;
+
+	node = bt_mesh_cdb_node_alloc(test_va_uuid, addr, comp.elem_count, 0);
+	ASSERT_TRUE(node);
+	memcpy(node->dev_key, dev_key, 16);
+}
+
+static void prov_and_conf_rx(uint16_t addr)
+{
+	provision(RX_ADDR);
+	common_configure(RX_ADDR);
+
+	target_node_alloc(rx_comp, TX_ADDR, dev_key);
+}
+
+static void test_priv_tx_init(void)
+{
+	provision(TX_ADDR);
+	common_configure(TX_ADDR);
+}
+
+static void wait_for_next_priv_beacon(uint8_t timeout)
```

Comment:
There is no difference in algorithms of working of SNB and PRB. The only difference is the fields and the rules that are used to generate them. Therefore, I guess `wait_for_beacon` can be used instead of this one. It can, for example, take beacon type as argument if needed. Or there can be a global variable storing what kind of beacon is tested now. If needed, this beacon struct can be extended to store PRB fields:
https://github.com/zephyrproject-rtos/zephyr/blob/a54e0bc98d82b6a76ffbedc5a9a5b3de6734c498/tests/bsim/bluetooth/mesh/src/test_beacon.c#L314-L319

Reply:
`wait_for_beacon` creates bt listener via `bt_le_scan_start`, which is not possible to use when Mesh is enabled on device, at least with current config. I think current approach, when RX device runs PRB Client that sets state on PRB Server makes sense. As far as I can see, `wait_for_beacon` is not used in any test that actually initiates Mesh. This looks like a reason why for example `test_rx_on_key_refresh` just calls:
```c
k_sleep(K_SECONDS(BEACON_INTERVAL));
```
instead of `wait_for_beacon`. In private beacon tests I just register additional scan callback that doesn't affect other, Mesh scan callback in particular. So all advertisements are passed here and to host.

Reply:
No, I'm currently reworking this PR, `wait_for_beacon` will be the only waiting function and as you said, in most cases we will be just waiting for `BEACON_INTERVAL`. This was introduced because of misunderstanding of what TX/RX devices should be setup as (TX device should be IUT sending Private Beacons, and RX device is control device with default composition that is not tested, but only listens/receives beacons). Because of this `wait_for_beacon` will be sufficient, maybe extended